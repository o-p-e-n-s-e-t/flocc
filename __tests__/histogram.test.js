const { Agent, Environment, Histogram, utils } = require("../dist/flocc");
const fs = require("fs");
const PNG = require("pngjs").PNG;
const pixelmatch = require("pixelmatch");
const puppeteer = require("puppeteer");
let browser, page;

const width = 200;
const height = 400;

const container = document.createElement("div");
const environment = new Environment();
let histogram;

beforeEach(() => {
  histogram = new Histogram(environment, {
    buckets: 4,
    width,
    height
  });
  histogram.metric("x");
  histogram.mount(container);
  histogram.canvas.getContext("2d").__clearEvents();
  histogram.canvas.getContext("2d").__clearDrawCalls();
});

afterEach(() => {
  while (environment.getAgents().length > 0) {
    environment.getAgents().pop();
  }
});

it("Correctly instantiates a histogram", () => {
  expect(histogram.canvas).toBeInstanceOf(HTMLCanvasElement);
  expect(histogram.background).toBeInstanceOf(HTMLCanvasElement);

  expect(histogram.environment).toBe(environment);
  expect(environment.renderers).toContain(histogram);

  expect(histogram.width).toEqual(width);
  expect(histogram.height).toEqual(height);
});

it("Correctly renders", () => {
  environment.tick();
  const calls = histogram.canvas.getContext("2d").__getDrawCalls();
  expect(calls).toMatchSnapshot();
});

it("Correctly renders with agents", () => {
  for (let i = 0; i < 100; i++) {
    const agent = new Agent();
    agent.set("x", Math.sqrt(i));
    environment.addAgent(agent);
  }
  environment.tick();
  const calls = histogram.canvas.getContext("2d").__getDrawCalls();
  expect(calls).toMatchSnapshot();
});

it("Correctly renders with discrete buckets", () => {
  const env = new Environment();
  const h = new Histogram(env, {
    background: 'yellow',
    buckets: [1, 2, 3],
    width,
    height
  });
  h.metric("x");
  h.mount(container);
  for (let i = 0; i < 10; i++) {
    const agent = new Agent({ x: i % 4 });
    env.addAgent(agent);
  }
  env.tick();
  expect(true).toBe(true);
  const calls = h.canvas.getContext("2d").__getDrawCalls();
  expect(calls).toMatchSnapshot();
});

it("Renders Flocking Histogram test correctly.", async () => {
  browser = await puppeteer.launch();
  page = await browser.newPage();
  try {
    await page.goto("http://localhost:3000/flocking-histogram", {
      waitUntil: "networkidle2"
    });
  } catch {
    console.warn(
      "Could not connect to localhost:3000, so skipping a Histogram test. Run `npm run dev` in a separate terminal window to make sure all tests run."
    );
    return await browser.close();
  }
  const filePath = __dirname + "/screenshots/flocking-histogram.png";
  const existingImage = fs.existsSync(filePath)
    ? PNG.sync.read(fs.readFileSync(filePath))
    : null;
  await page.screenshot({ path: filePath });
  if (!existingImage) {
    return await browser.close();
  }
  const { width, height } = existingImage;
  const newImage = PNG.sync.read(fs.readFileSync(filePath));
  const diff = new PNG({ width, height });
  expect(
    pixelmatch(existingImage.data, newImage.data, diff.data, width, height)
  ).toBe(0);

  await browser.close();
})