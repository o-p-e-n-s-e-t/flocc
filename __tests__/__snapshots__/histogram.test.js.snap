// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Correctly renders 1`] = `
Array [
  Object {
    "props": Object {
      "height": 400,
      "width": 200,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "0...0.25",
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      0.7071067811865476,
      0.7071067811865475,
      -0.7071067811865475,
      0.7071067811865476,
      38.75,
      350,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "0.25...0.5",
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      0.7071067811865476,
      0.7071067811865475,
      -0.7071067811865475,
      0.7071067811865476,
      76.25,
      350,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "0.5...0.75",
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      0.7071067811865476,
      0.7071067811865475,
      -0.7071067811865475,
      0.7071067811865476,
      113.75,
      350,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "0.75...1",
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      0.7071067811865476,
      0.7071067811865475,
      -0.7071067811865475,
      0.7071067811865476,
      151.25,
      350,
    ],
    "type": "fillText",
  },
]
`;

exports[`Correctly renders with agents 1`] = `
Array [
  Object {
    "props": Object {
      "height": 400,
      "width": 200,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "0",
      "x": 3,
      "y": 340,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 13,
            "y": 340,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": 200,
            "y": 340,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "20",
      "x": 3,
      "y": 272,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 13,
            "y": 272,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": 200,
            "y": 272,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "40",
      "x": 3,
      "y": 204,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 13,
            "y": 204,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": 200,
            "y": 204,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "60",
      "x": 3,
      "y": 136,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 13,
            "y": 136,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": 200,
            "y": 136,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "80",
      "x": 3,
      "y": 68,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 13,
            "y": 68,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": 200,
            "y": 68,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "100",
      "x": 3,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 13,
            "y": 0,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": 200,
            "y": 0,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "0...0.25",
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      0.7071067811865476,
      0.7071067811865475,
      -0.7071067811865475,
      0.7071067811865476,
      41.375,
      350,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "0.25...0.5",
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      0.7071067811865476,
      0.7071067811865475,
      -0.7071067811865475,
      0.7071067811865476,
      78.125,
      350,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "0.5...0.75",
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      0.7071067811865476,
      0.7071067811865475,
      -0.7071067811865475,
      0.7071067811865476,
      114.875,
      350,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "0.75...1",
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      0.7071067811865476,
      0.7071067811865475,
      -0.7071067811865475,
      0.7071067811865476,
      151.625,
      350,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "height": 3.4,
      "width": 29.400000000000002,
      "x": 26.675,
      "y": 336.6,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 0,
      "width": 29.400000000000002,
      "x": 63.425000000000004,
      "y": 340,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 0,
      "width": 29.400000000000002,
      "x": 100.175,
      "y": 340,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 3.4,
      "width": 29.400000000000002,
      "x": 136.925,
      "y": 336.6,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
]
`;

exports[`Correctly renders with discrete buckets 1`] = `
Array [
  Object {
    "props": Object {
      "height": 400,
      "width": 200,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "0",
      "x": 2,
      "y": 340,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 12,
            "y": 340,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": 200,
            "y": 340,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "2",
      "x": 2,
      "y": 272,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 12,
            "y": 272,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": 200,
            "y": 272,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "4",
      "x": 2,
      "y": 204,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 12,
            "y": 204,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": 200,
            "y": 204,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "6",
      "x": 2,
      "y": 136,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 12,
            "y": 136,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": 200,
            "y": 136,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "8",
      "x": 2,
      "y": 68,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 12,
            "y": 68,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": 200,
            "y": 68,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "10",
      "x": 2,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 12,
            "y": 0,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": 200,
            "y": 0,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "1",
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      0.7071067811865476,
      0.7071067811865475,
      -0.7071067811865475,
      0.7071067811865476,
      46.66666666666667,
      350,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "2",
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      0.7071067811865476,
      0.7071067811865475,
      -0.7071067811865475,
      0.7071067811865476,
      96,
      350,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "3",
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      0.7071067811865476,
      0.7071067811865475,
      -0.7071067811865475,
      0.7071067811865476,
      145.33333333333334,
      350,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "height": 102,
      "width": 39.46666666666667,
      "x": 26.933333333333334,
      "y": 238,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 68,
      "width": 39.46666666666667,
      "x": 76.26666666666668,
      "y": 272,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 68,
      "width": 39.46666666666667,
      "x": 125.6,
      "y": 272,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
]
`;
